File: java/weak_typing/Tournament.java
import java.util.Comparator; import java.util.Random; import java.util.Arrays;

class Tournament
{
  Comparator<Object> comparator; Object tourney;

  Tournament( Comparator<Object> comparator, Object[] v )
  {
    this.comparator = comparator;
    this.tourney = knockout( v, 0, v.length-1 );
  }

  public Object pop()
  {
    Object result = Match.winner( tourney );
    tourney = tourney instanceof Match ? ((Match)tourney).remove_winner( comparator ) : null; return result;
  }

  public static void sort( Object[] v, Comparator<Object> comparator )
  {
    Tournament tourney = new Tournament( comparator, v );
    for (int i = 0; i < v.length; i++)
      v[i] = tourney.pop();
    }

  private Object knockout( Object[] v, int i, int k )
  {
    if (i == k) return v[i];
    int j = (i+k)/2;
    return new Match( comparator, knockout( v, i, j ), knockout( v, j+1, k ) );
  }

  static class Match
  {
    Object losers, winners, item;

    Match( Comparator<Object> comparator, Object a, Object b )
    {
      Object a_winner = winner( a ), b_winner = winner( b );
      boolean result = comparator.compare( a_winner, b_winner ) < 0;
      this.losers = result ? b : a;
      this.winners = result ? a : b;
      this.item = result ? a_winner : b_winner;
    }

    static Object winner( Object t )
    {
      return t instanceof Match ? ((Match)t).item : t;
    }

    public String toString() { return "([" + item + "]" + winners + losers + ")"; }

    public Object remove_winner( Comparator<Object> comparator )
    {
      return winners instanceof Match ? new Match( comparator, losers, ((Match)winners).remove_winner( comparator ) ) : losers;
    }
  }

  static Integer[] randomInts( int n )
  {
    Random r = new Random( 12345678 );
    Integer[] v = new Integer[n];
    for (int i = 0; i < n; i++)
      v[i] = r.nextInt( 10*n );
    return v;
  }

  static void time( String description, Runnable action, InstrumentedCompare compare )
  {
    long start = System.currentTimeMillis();
    action.run();
    long finish = System.currentTimeMillis();
    System.out.println( description + " took " + ((double)(finish-start)/1000.0) + " secs " + compare.count + " comparisons");
  }

  static class InstrumentedCompare implements Comparator<Object>
  {
    public int count = 0;  
    public int compare( Object a, Object b ) { count++; return ((Integer)a).compareTo( (Integer)b ); }
  }

  public static void main( String[] args )
  {
    int n = args.length >= 1 ? Integer.parseInt( args[0] ) : 100000;
    final InstrumentedCompare tournamentCompare = new InstrumentedCompare();
    final Integer[] nums = randomInts( n );
    time( "Tournament.sort", new Runnable(){ public void run(){ Tournament.sort( nums, tournamentCompare ); } }, tournamentCompare );
    final Integer[] nums2 = randomInts( n );
    final InstrumentedCompare systemCompare = new InstrumentedCompare();
    time( "Array.sort", new Runnable(){ public void run(){ Arrays.sort( nums2, systemCompare ); } }, systemCompare );
    for (int i = 0; i < n; i++)
      if (nums[i].compareTo( nums2[i] ) != 0)
      {
        System.err.println( "Arrays do not match at index: " + i );
        return;
      }
  }
}

File: ruby/weak_typing/tournament.rb
class Match
  attr_reader :winner, :losers, :winners
  @@count = 0
  
  def initialize( top, bot )
    @@count += 1
    top_winner, bot_winner = Match.get_winner( top ), Match.get_winner( bot )
    @winner, @winners, @losers = top_winner <= bot_winner ? [top_winner, top, bot] : [bot_winner, bot, top]
  end

  def to_s
    "([#{@winner}] #{@winners.to_s} #{@losers.to_s})"
  end
  
  def Match.get_winner( t )
    t.is_a?( Match ) ? t.winner : t
  end
  
  def Match.extract_winner( t )
    t.is_a?( Match ) ? [t.winner, t.remove_winner] : [t, nil]
  end
  
  def remove_winner
    winners.is_a?( Match ) ? Match.new( losers, winners.remove_winner ) : losers;
  end
  
  def Match.count
    @@count
  end
end

class Tournament
  def initialize( v )
    @tourney = knockout( v )
  end
  
  def pop
    winner, @tourney = Match.extract_winner( @tourney )
    winner
  end
  
  def knockout( v, i=0, k=v.size-1 )
    i == k ? v[i] : (j = (i+k)/2; Match.new( knockout( v, i, j ), knockout( v, j+1, k ) ))
  end
  
  def self.sort( v )
    tourney = Tournament.new( v )
    v.size.times{ |i| v[i] = tourney.pop }
  end
end

def log2( n )
  Math.log( n )/Math.log( 2 )
end

def log2fact( n )
  2.upto( n ).inject( 0 ){ |sum,i| sum + log2( i ) }.ceil
end

srand( 12345678 )
n = (ARGV.size >= 1 && ARGV[0].to_i) || 1024
m = n*n
test = Array.new( n ){ |i| rand( m ) }
puts test.inspect if test.size <= 64
start = Time.now
sorted = Tournament.sort( test )
elapsed = Time.now - start
puts sorted.sorted.inspect if test.size <= 64
puts Match::count.to_f/log2fact( n )
puts "#{elapsed} secs"

test = Array.new( n ){ |i| rand( m ) }
start = Time.now
sorted = test.sort
elapsed = Time.now - start
puts "#{elapsed} secs"